libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2228,18 +2236,23 @@ string YulUtilFunctions::readFromMemoryOrCalldata(Type const& _type, bool _fromC
 				("splitFunction", splitExternalFunctionIdFunction())
 				.render();
 
+
 		return Whiskers(R"(
-			function <functionName>(memPtr) -> value {
-				value := <load>(memPtr)
-				<?needsValidation>
+			function <functionName>(ptr) -> value {
+				<?fromCalldata>
+					value := calldataload(ptr)
 					<validate>(value)
-				</needsValidation>
+				<!fromCalldata>
+					value := <cleanup>(mload(ptr))
+				</fromCalldata>
 			}
 		)")
 		("functionName", functionName)
-		("load", _fromCalldata ? "calldataload" : "mload")
-		("needsValidation", _fromCalldata)
-		("validate", _fromCalldata ? validatorFunction(_type) : "")
+		("fromCalldata", _fromCalldata)
+		("validate", validatorFunction(_type))
+		// Apart from reading byte array elements, we primarily clean up anticipating dirty memory writes
+		// form inline assembly. TODO: Rethink that.
```

Comment:
`Rethink` as in, if this should be done at all, or how to do it?

Reply:
If it should be done at all - I'll rephrase the comment, though. We'll probably keep doing it, but I'll retain the comment about why.
I wonder how well the optimizer will ever be able to remove the cleanups for these, though...